### ISSUE TITLE
Using Swing and AsyncS3Client to make a dowload of a directory stops after 1st file.

### ISSUE THREAD

**Issue by studioam** [aws/aws-sdk-java-v2#1221]
I'm trying to create a download directory tool using a simple Swing GUI with a ProgressBar and an 
AsyncS3Client. I'm using an AsyncS3Client because a Sync client works fine (already tested) but blocks everything including the swing GUI and the progressbar.  I have to perform the download of a directory containing at least 5 files of 30GB each.

this is my HttpAsync Client:

    SdkAsyncHttpClient httpClientBuilder = NettyNioAsyncHttpClient.builder().connectionTimeout(connectionTimeout).maxConcurrency(80).maxPendingConnectionAcquires(10_000).build();

and the  s3client 

    s3 = S3AsyncClient.builder().region(Region.EU_WEST_1).credentialsProvider(credProv).httpClient(httpClientBuilder).build();

Assuming the process might fail for any reason, I want the files to be downloaded one at a time, and a file download should start only after the previous one has succesfully completed. In case of failure the tool checks size of each file and if there's a truncated one it gets deleted and download restarts from that file.

So before starting the global download process I make put every file name and size in an HashMap where name is key and size is value. The HashMap is my download list, and in case of failure it gets purged of succesfully completed files upon restarting the program.

this is the Download Cycle code:

    for (HashMap.Entry<String, Long> s: DCPFiles.entrySet()) {
                
                    while (!(s3.getObject(GetObjectRequest.builder().bucket(bucketName).key(s.getKey()).build(),
                       AsyncResponseTransformer.toFile(Paths.get(path.getPath()+ "\\" + s.getKey()))).isDone())){
                    
                                 
                 
                    
                 jLabel5.setText("Download file " + Integer.toString(directory.listFiles().length) + " di " + Integer.toString(nFilesDCP) + " del DCP    "+ DCPname);
                
                }   
             }

it works fine with the 1st file, Swing GUI gets acknowledged but when the file download ends it will not continue with the next. Can anyone explain Why? Is it required something like a nextContinuationToken ?

With Sync the GUI is frozen, but the download goes perfectly

**Comment by varunnvs92** [comment]
You are calling `s3.getObject` on each while call, so a new download will start each time the while condition is executed. You need to move the getObject call outside while loop, something like:
```
        CompletableFuture<ReturnType> future = 
            s3.getObject(GetObjectRequest.builder().bucket(bucketName).key(s.getKey()).build(),
                         AsyncResponseTransformer.toFile(Paths.get(path.getPath() + "\\" + s.getKey())));
       
        while (!future.isDone()) {
              jLabel5.setText("Downloa...
        }
```

I would suggest only raising issues for Bugs and Feature requests as we have limited bandwidth. For `How To` questions, we recommend reaching out to developer community platforms like `StackOverflow` and tag it with `aws-java-sdk`. Thank you.

**Comment by studioam** [comment]
OK, thanks. Will keep in mind for future requests.
About your kind reply, and staying within possible bugs/behaviour reports, your solution works fine but leads to the same result as I have with Sync server, meaning the GUI freezes and jLabel5.setText("Download...              isn't executed. Both progress bar and label content won't get acknowledged during download.

**Comment by varunnvs92** [comment]
Where is the application blocking? `isDone()` method is non-blocking and so the code inside while loop should execute. As there is no `Thread.Sleep` inside while loop, the statement `jLabel5.setText(` gets executed very frequently until download is done. Maybe it is causing your application to hang (as you are calling setText very frequently) but I am not familiar with Swing and can't give good suggestions there.

Basically as per your code, the execution stays inside while loop until the download is done. If that is not what you want, you need to run that code in background thread.

**Comment by studioam** [comment]
Thanks for your reply, I assumed so, but in fact when you hit "Start" the button remains freezed and everything in the Swing GUI is blocked, the program is not stuck because the download goes on correctly, but the upper layer caller is blocked until isDone() gets completed I figure out. And the text in the label doesn't appear. It's exactly as it happens if using a Sync client. Please look at the picture here: [https://postimg.cc/34DNRHJR](url)

**Comment by dagnir** [comment]
Hi, as @varunnvs92 has pointed out, the issue is that the GUI thread is blocked on the `while` loop, which is polling the `isDone` method; the way to avoid that would be update the download status asynchronously using a separate thread so the GUI thread is not blocked.

As this seems like a Swing specific question rather than SDK one, I'm going to go ahead and close this. I would suggest asking the question over at StackOverflow where users more familiar with Swing would be able to help you out.
